File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/service/StatusService.java
package br.com.fiap.bluegather.service;

import br.com.fiap.bluegather.dto.StatusDTO;
import br.com.fiap.bluegather.model.Status;
import br.com.fiap.bluegather.repository.StatusRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.Objects;

@Service
public class StatusService {

    @Autowired
    private StatusRepository statusRepository;

    public Page<StatusDTO> listAll(Pageable pageRequest) {
        Page<Status> list = statusRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public StatusDTO findById(Long id) {
        Status entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public StatusDTO create(StatusDTO newData) {
        Status entity = convertToEntity(newData);
        Status savedEntity = statusRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public StatusDTO update(Long id, StatusDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Status updatedEntity = convertToEntity(updatedData);    
        Status savedEntity = statusRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Status entity = findEntityById(id);
        statusRepository.delete(entity);
    }

    public Status findEntityById(Long id) {
        return statusRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Status não encontrado(a) por ID: " + id));
    }
    
    public StatusDTO convertToDto(Status entity) {
        StatusDTO dto = new StatusDTO();
        dto.setId(entity.getId());
        dto.setNome(entity.getNome());
        return dto;
    }

    public Status convertToEntity(StatusDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - StatusDTO não pode ser nulo.");
        }
        Status entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Status();
        }
        entity.setNome(dto.getNome());
        return entity;
    }
}

File: Enterprise Application Development/bluegather/Services/StatusService.cs
using Microsoft.EntityFrameworkCore;
using BlueGather.Dto;
using BlueGather.Models;
using BlueGather.Context;

namespace BlueGather.Services;
public class StatusService
{
    private readonly BlueGatherContext _context;

    public StatusService(BlueGatherContext context)
    {
        _context = context;
    }

    public async Task<List<StatusDto>> FindAll()
    {
        var list = await _context.Status
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<StatusDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<StatusDto> Create(StatusDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Status.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<StatusDto> Update(long id, StatusDto updatedData)
    {
        var entity = await FindEntityById(id);
        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Status.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<StatusModel> FindEntityById(long id)
    {
        var entity = await _context.Status
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<StatusDto>> FindByName(string nome)
    {
        var list = await _context.Status
            .Where(x => x.Nome.ToLower().Contains(nome.ToLower()))
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private StatusDto ConvertToDto(StatusModel entity)
    {
        return new StatusDto
        {
            Id = entity.Id,
            Nome = entity.Nome,
        };
    }

    private async Task<StatusModel> ConvertToEntity(StatusDto dto)
    {
        return new StatusModel
        {
            Id = dto.Id ?? 0,
            Nome = dto.Nome
        };
    }
}